src/trajectory.jl

```diff
@@ -82,121 +141,200 @@ end
+        sampler = BaseSampler(sampler, H′)
+        s′ = iscontinued(sampler, nt, H, H′)
         α′ = exp(min(0, H - H′))
-
-        return z′, z′, z′, n′, s′, α′, 1
+        return DoublingTree(z′, z′, z′, sampler, s′, α′, 1)
     else
         # Recursion - build the left and right subtrees.
-        zm, zp, z′, n′, s′, α′, n′α = build_tree(rng, nt, h, z, logu, v, j - 1, H)
-
-        if s′ == 1
+        dt′ = build_tree(rng, nt, h, z, sampler, v, j - 1, H)
+        # Expand tree if not terminated
+        if dt′.s == 1
+            # Expand left
             if v == -1
-                zm, _, z′′, n′′, s′′, α′′, n′′α = build_tree(rng, nt, h, zm, logu, v, j - 1, H)
-            else
-                _, zp, z′′, n′′, s′′, α′′, n′′α = build_tree(rng, nt, h, zp, logu, v, j - 1, H)
-            end
-            if rand(rng) < n′′ / (n′ + n′′)
-                z′ = z′′
+                dt′′ = build_tree(rng, nt, h, dt′.zleft, sampler, v, j - 1, H) # left tree
+                dtleft, dtright = dt′′, dt′
+            # Expand right
+            else    
+                dt′′ = build_tree(rng, nt, h, dt′.zright, sampler, v, j - 1, H) # right tree
+                dtleft, dtright = dt′, dt′′
             end
-            α′ = α′ + α′′
-            n′α = n′α + n′′α
-            s′ = s′′ * (dot(zp.θ - zm.θ, ∂H∂r(h, zm.r)) >= 0 ? 1 : 0) * (dot(zp.θ - zm.θ, ∂H∂r(h, zp.r)) >= 0 ? 1 : 0)
-            n′ = n′ + n′′
+            dt′ = merge(rng, h, dtleft, dtright)
         end
-
-        # s: termination stats
-        # α: MH stats, i.e. sum of MH accept prob for all leapfrog steps
-        # nα: total # of leap frog steps, i.e. phase points in a trajectory
-        # n: # of acceptable candicates, i.e. prob is larger than slice variable u
-        return zm, zp, z′, n′, s′, α′, n′α
+        return dt′
     end
 end
 
+"""
+Recursivly build a tree for a given depth `j`.
+"""
 build_tree(
-    nt::DynamicTrajectory{I},
+    nt::NUTS{I,F,S},
     h::Hamiltonian,
     z::PhasePoint,
-    logu::AbstractFloat,
+    sampler::AbstractNUTSSampler,
     v::Int,
     j::Int,
     H::AbstractFloat
-) where {I<:AbstractIntegrator,T<:Real} = build_tree(GLOBAL_RNG, nt, h, z, logu, v, j, H)
+) where {I<:AbstractIntegrator,F<:AbstractFloat,S<:AbstractNUTSSampling} = build_tree(GLOBAL_RNG, nt, h, z, sampler, v, j, H)
+
+mh_accept(rng::AbstractRNG, s::SliceNUTSSampler, s′::SliceNUTSSampler) = mh_accept(rng, s.n, s′.n)
+mh_accept(rng::AbstractRNG, s::MultinomialNUTSSampler, s′::MultinomialNUTSSampler) = mh_accept(rng, s.w, s′.w)
+mh_accept(s::AbstractNUTSSampler, s′::AbstractNUTSSampler) = mh_accept(GLOBAL_RNG, s, s′)
+
+InitSampler(rng::AbstractRNG, ::SliceNUTSSampling, H::AbstractFloat) = SliceNUTSSampler(log(rand(rng)) - H, 1)
+InitSampler(rng::AbstractRNG, ::MultinomialNUTSSampling, H::AbstractFloat) = MultinomialNUTSSampler(H)
+InitSampler(s::AbstractNUTSSampling, H::AbstractFloat) = InitSampler(GLOBAL_RNG, s, H)
 
 function transition(
```

Comment:
@yebai So here I didn't use `FullBinaryTree` to initialize the starting point. 
- The main reason being the different sampling steps: in `build_tree` sampling is done by the sampler while here is done by a MH step. 
- Additionally the conditon to do the sampling are different: here MH step is only done if the proposed tree is not terminated while that in `build_tree` is performed regardless (maybe this is something we can optimize?). 
- For this two reasons I didn't find a good way to unify the `merge` so that it can also be used here.
- Also I don't want to return the tree anyway so there might be less necessity to do so.